src/Lean/Elab/Tactic/BvDecide/BitBlast/BVExpr/BitBlast/Lemmas/Add.lean

```diff
@@ -0,0 +1,264 @@
+          =
+        (s.get idx hidx).cast (by apply go_le_size) := by
+  intros
+  apply go_get_aux
+
+theorem atLeastTwo_eq_halfAdder (lhsBit rhsBit carry : Bool) :
+    Bool.atLeastTwo lhsBit rhsBit carry
+      =
+    (((xor lhsBit rhsBit) && carry) || (lhsBit && rhsBit)) := by
+  revert lhsBit rhsBit carry
+  decide
+
+theorem go_denote_eq (aig : AIG α) (curr : Nat) (hcurr : curr ≤ w) (cin : Ref aig)
+    (s : AIG.RefVec aig curr) (lhs rhs : AIG.RefVec aig w) (assign : α → Bool)
+    (lhsExpr rhsExpr : BitVec w)
+    (hleft : ∀ (idx : Nat) (hidx : idx < w), ⟦aig, lhs.get idx hidx, assign⟧ = lhsExpr.getLsb idx)
+    (hright : ∀ (idx : Nat) (hidx : idx < w), ⟦aig, rhs.get idx hidx, assign⟧ = rhsExpr.getLsb idx)
+    (hcin : ⟦aig, cin, assign⟧ = BitVec.carry curr lhsExpr rhsExpr false) :
+    ∀ (idx : Nat) (hidx1 : idx < w),
+        curr ≤ idx
+          →
+        ⟦
+          (go aig lhs rhs curr hcurr cin s).aig,
+          (go aig lhs rhs curr hcurr cin s).vec.get idx hidx1,
+          assign
+        ⟧
+          =
+        ⟦aig, lhs.get idx hidx1, assign⟧.xor
+          (⟦aig, rhs.get idx hidx1, assign⟧.xor
+            (BitVec.carry idx lhsExpr rhsExpr false)) := by
+  intro idx hidx1 hidx2
+  generalize hgo : go aig lhs rhs curr hcurr cin s = res
+  unfold go at hgo
+  dsimp only at hgo
+  split at hgo
+  . next hlt =>
+    cases Nat.eq_or_lt_of_le hidx2 with
+    | inl heq =>
+      rw [← hgo]
+      rw [go_get (hidx := by omega)]
+      rw [AIG.RefVec.get_push_ref_eq' (hidx := by rw [heq])]
+      simp only [← heq]
+      rw [go_denote_mem_prefix]
+      . unfold mkFullAdder
+        simp [hcin]
+      . simp only [Ref_cast']
+        apply Ref.hgate
+    | inr hlt =>
+      rw [← hgo]
+      rw [go_denote_eq (lhsExpr := lhsExpr) (rhsExpr := rhsExpr) (curr := curr + 1)]
+      . rw [mkFullAdder_denote_mem_prefix]
+        rw [mkFullAdder_denote_mem_prefix]
+        . simp
+        . simp [Ref.hgate]
+        . simp [Ref.hgate]
+      . intro idx hidx
+        rw [mkFullAdder_denote_mem_prefix]
+        rw [← hleft idx hidx]
+        . simp
+        . simp [Ref.hgate]
+      . intro idx hidx
+        rw [mkFullAdder_denote_mem_prefix]
+        rw [← hright idx hidx]
+        . simp
+        . simp [Ref.hgate]
+      . unfold mkFullAdder
+        simp only [Ref_cast', id_eq, Int.reduceNeg, denote_projected_entry, denote_mkFullAdderCarry,
+          FullAdderInput.lhs_cast, FullAdderInput.rhs_cast, FullAdderInput.cin_cast,
+          BitVec.carry_succ]
+        rw [AIG.LawfulOperator.denote_mem_prefix (f := mkFullAdderOut)]
+        rw [AIG.LawfulOperator.denote_mem_prefix (f := mkFullAdderOut)]
+        rw [AIG.LawfulOperator.denote_mem_prefix (f := mkFullAdderOut)]
+        rw [hleft, hright, hcin]
+        simp [atLeastTwo_eq_halfAdder]
+      . omega
+  . omega
+termination_by w - curr
+
+end blastAdd
+
+theorem blastAdd_denote_eq (aig : AIG α) (lhs rhs : BitVec w) (assign : α → Bool)
```

Comment:
Technically many theorems here have incorrect names, for example this one should be called `denote_blastAdd`.